src/engine/memory.rs

```diff
@@ -0,0 +1,113 @@
+use std::ops::{Index, IndexMut};
+
+#[derive(Debug, Clone)]
+pub struct Memory<T> {
+    memory_size: usize,
+    segment_mask: usize,
+    segment_shift: u32,
+    default: T,
```

Comment:
Rust has incredible support for default values. It's more idiomatic to require T to implement Default<T>, such that you can create default values with T::default().

Reply:
Done. Note that I still had to keep the `default` field in the struct because the (non-mutable) `Index` accessor needs to return a reference to the default value (not the value itself). But I am now initializing the field via `T::default()` during construction.